packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -227,6 +240,107 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
+     */
+    function closeShortWithUserNft(CloseShortWithUserNftParams calldata params) external {
+        INonfungiblePositionManager(nonfungiblePositionManager).safeTransferFrom(
+            msg.sender,
+            address(this),
+            params.tokenId
+        );
+
+        // get liquidity amount, and withdraw ETH and wPowerPerp amounts in LP position
+        (uint128 liquidity, , ) = ControllerHelperLib._getUniPositionBalances(
+            nonfungiblePositionManager,
+            params.tokenId,
+            IOracle(oracle).getTimeWeightedAverageTickSafe(wPowerPerpPool, 420),
+            isWethToken0
+        );
+        // (
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     uint128 liquidity,
+        //     ,
+        //     ,
+        //     ,
+        // ) = INonfungiblePositionManager(nonfungiblePositionManager).positions(params.tokenId);
+        INonfungiblePositionManager.DecreaseLiquidityParams memory decreaseParams = INonfungiblePositionManager
+            .DecreaseLiquidityParams({
+                tokenId: params.tokenId,
+                liquidity: liquidity,
+                amount0Min: params.amount0Min,
+                amount1Min: params.amount1Min,
+                deadline: block.timestamp
+            });
+        uint256 wethAmount;
+        uint256 wPowerPerpAmount;
+        (isWethToken0)
+            ? (wethAmount, wPowerPerpAmount) = INonfungiblePositionManager(nonfungiblePositionManager)
+                .decreaseLiquidity(decreaseParams)
+            : (wPowerPerpAmount, wethAmount) = INonfungiblePositionManager(nonfungiblePositionManager)
+            .decreaseLiquidity(decreaseParams);
+        (uint256 amount0, uint256 amount1) = INonfungiblePositionManager(nonfungiblePositionManager).collect(
+            INonfungiblePositionManager.CollectParams({
+                tokenId: params.tokenId,
+                recipient: address(this),
+                amount0Max: isWethToken0 ? uint128(wethAmount) : uint128(wPowerPerpAmount),
+                amount1Max: isWethToken0 ? uint128(wPowerPerpAmount) : uint128(wethAmount)
+            })
+        );
+
+        if (wPowerPerpAmount < params.wPowerPerpAmountToBurn) {
+            // swap needed wPowerPerp amount to close short position
+            // TODO: need to set max slippage here
+            _exactOutFlashSwap(
+                weth,
+                wPowerPerp,
+                IUniswapV3Pool(wPowerPerpPool).fee(),
+                params.wPowerPerpAmountToBurn.sub(wPowerPerpAmount),
+                IWETH9(weth).balanceOf(address(this)),
+                uint8(FLASH_SOURCE.SWAP_EXACTOUT_ETH_WPOWERPERP),
+                ""
+            );
+
+            IController(controller).burnWPowerPerpAmount(
+                params.vaultId,
+                params.wPowerPerpAmountToBurn,
+                params.collateralToWithdraw
+            );
+        } else {
```

Comment:
If both values are equal, it will just burn without swapping https://github.com/opynfinance/squeeth-monorepo/pull/210/files#diff-f190257c33d00ce0e20f170faf26f3057272439fa7cefa812661b53f1b58badcR324 (L324)

Reply:
ah yeah see ln333 you check the excess